promql/prettier/prettier_test.go

```diff
@@ -0,0 +1,966 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prettier
+
+import (
+	"testing"
+
+	"github.com/prometheus/prometheus/util/testutil"
+)
+
+type prettierTest struct {
+	expr     string
+	expected string
+}
+
+var sortingLexItemCases = []prettierTest{
+	{
+		expr:     `sum(go_alloc_bytes) by (job)`,
+		expected: `sum by (job) (go_alloc_bytes)`,
+	},
+	{
+		expr:     `sum by (job) (go_alloc_bytes)`,
+		expected: `sum by(job) (go_alloc_bytes)`,
+	},
+	{
+		expr:     `metric_first + metric_second`,
+		expected: `metric_first + metric_second`,
+	},
+	{
+		expr:     `metric_first + bool metric_second`,
+		expected: `metric_first + bool metric_second`,
+	},
+	{
+		expr:     `metric_first + ignoring(job) metric_second`,
+		expected: `metric_first + ignoring(job) metric_second`,
+	},
+	{
+		expr:     `sum without(job, instance, foo) (metric_first + metric_second)`,
+		expected: `sum without(job, instance, foo) (metric_first + metric_second)`,
+	},
+	{
+		expr:     `quantile(0.9, sum(go_goroutines) without(job)) by (localhost)`,
+		expected: `quantile by (localhost) (0.9, sum without(job) (go_goroutines))`,
+	},
+	{
+		expr:     `quantile(0.9, sum(min(go_alloc_bytes) by (job)) without(job)) by (localhost)`,
+		expected: `quantile by (localhost) (0.9, sum without(job) (min by (job) (go_alloc_bytes)))`,
+	},
+	{
+		expr:     `quantile(0.9, sum(min(go_alloc_bytes) by (job)) without(job)) by (localhost)`,
+		expected: `quantile by(localhost) (0.9, sum without(job) (min by(job) (go_alloc_bytes)))`,
+	},
+	{
+		expr:     `quantile(0.9, sum(min(go_alloc_bytes) by (job)) ignoring(job)) by (localhost)`,
+		expected: `quantile by (localhost) (0.9, sum ignoring(job) (min by (job) (go_alloc_bytes)))`,
```

Comment:
It'd be too to see how this looks with the full prettier.

Reply:
This is the prettified output. I will add the same as a unit-test.

`quantile by(instance) (0.9, sum without(job) (min by(job) (go_alloc_bytes)))`

Reply:
That's very dense semantically, you could easily miss the nonsense aggregation here. I think there shouldn't be more than one aggregation per line.

Reply:
Addressed. Now, if any aggregate expression has an aggregate expression as an immediate parent, it will be printed on a separate line.